cache/cache.go

```diff
@@ -107,3 +117,95 @@ func (m *MemcachedCache) Set(k string, v []byte, expire int32) {
 func (m *MemcachedCache) Timeouts() uint64 {
 	return atomic.LoadUint64(&m.timeouts)
 }
+
+type ReplicatedMemcached struct {
+	prefix    string
+	instances []*memcache.Client
+
+	// was 50 ms
+	timeoutMs uint64 // timeout for getting data
+	// TODO (grzkv) Scrape this metric
+	timeouts uint64 // timeouts counter
+}
+
+// NewReplicatedMemcached creates a set of identical memcached instances.
+func NewReplicatedMemcached(prefix string, timeout uint64, servers ...string) BytesCache {
+	m := ReplicatedMemcached{
+		prefix:    prefix,
+		timeoutMs: timeout,
+	}
+
+	for _, s := range servers {
+		m.instances = append(m.instances, memcache.New(s))
+	}
+
+	return &m
+}
+
+// Get gets value for the key from the replicated memcached.
+// It sends the request to all replicas and picks the first valid answer
+// (event if it's a not-found) or times out.
+func (m *ReplicatedMemcached) Get(k string) ([]byte, error) {
+	resCh := make(chan cacheResponse)
+
+	tout := time.After(time.Duration(m.timeoutMs) * time.Millisecond)
+	cacheErrs := ""
+	for i := 0; i < len(m.instances); i++ {
+		select {
+		case res := <-resCh:
+			if res.err != nil {
+				cacheErrs = cacheErrs + "; " + res.err.Error()
+			} else if !res.found {
+				return nil, ErrNotFound
+			}
+
+			return res.data, nil
+		case <-tout:
+			atomic.AddUint64(&m.timeouts, 1)
+			return nil, ErrTimeout
+		}
+	}
+
+	// if this point is reached, it means that all caches returned errors
+	return nil, errors.New("all caches failed with errors: " + cacheErrs)
+}
+
+// Set sets the key-value pair for all cache instances.
+func (rm *ReplicatedMemcached) Set(k string, val []byte, expire int32) {
+	key := sha1.Sum([]byte(k))
+	hk := hex.EncodeToString(key[:])
+	for _, m := range rm.instances {
+		go func(k_ string, val_ []byte, expire_ int32, m_ *memcache.Client) {
+			m_.Set(&memcache.Item{
```

Comment:
We could add counter here for number of Set errors.

Reply:
This is a broader issue. We need to add a while set of metrics. In addition to that we also need to change the way the counting is done for old memcached. Currently we use an expvar where Prometheus should’ve been used. I will tackle that in a separate PR.